cmd/serverless-init/metadata/metadata.go

```diff
@@ -6,43 +6,123 @@
 package metadata
 
 import (
+	"fmt"
 	"io/ioutil"
 	"net/http"
+	"strings"
+	"sync"
 	"time"
 
+	"github.com/DataDog/datadog-agent/cmd/serverless-init/timing"
 	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
-const defaultURL = "http://metadata.google.internal/computeMetadata/v1/instance/id"
+const defaultBaseURL = "http://metadata.google.internal/computeMetadata/v1"
+const defaultContainerIDURL = "/instance/id"
+const defaultRegionURL = "/instance/region"
+const defaultProjectID = "/project/project-id"
 const defaultTimeout = 300 * time.Millisecond
 
-type Config struct {
-	url     string
-	timeout time.Duration
+type config struct {
+	containerIDURL string
+	regionURL      string
+	projectIDURL   string
+	timeout        time.Duration
 }
 
-func GetDefaultConfig() *Config {
-	return &Config{
-		url:     defaultURL,
-		timeout: defaultTimeout,
+type info struct {
+	tagName string
+	value   string
+}
+
+type Metadata struct {
+	containerID *info
+	region      *info
+	projectID   *info
+}
+
+func (metadata *Metadata) TagMap() map[string]string {
+	tagMap := map[string]string{}
+	if metadata.containerID != nil {
+		tagMap[metadata.containerID.tagName] = metadata.containerID.value
+	}
+	if metadata.region != nil {
+		tagMap[metadata.region.tagName] = metadata.region.value
+	}
+	if metadata.projectID != nil {
+		tagMap[metadata.projectID.tagName] = metadata.projectID.value
+	}
+	return tagMap
+}
+
+func GetDefaultConfig() *config {
+	return &config{
+		containerIDURL: fmt.Sprintf("%s%s", defaultBaseURL, defaultContainerIDURL),
+		regionURL:      fmt.Sprintf("%s%s", defaultBaseURL, defaultRegionURL),
+		projectIDURL:   fmt.Sprintf("%s%s", defaultBaseURL, defaultProjectID),
+		timeout:        defaultTimeout,
+	}
+}
+
+func GetMetaData(config *config) *Metadata {
+	wg := sync.WaitGroup{}
+	metadata := &Metadata{}
+	wg.Add(3)
+	go func() {
```

Comment:
Just to confirm, there is no way of getting all these values in a single API call?

Reply:
that's correct: https://cloud.google.com/run/docs/container-contract#metadata-server